datasets/dataset_plugin.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Callable, Dict, Iterable, Optional, Union
+
+from datasets.context import Context
+
+from .mode import Mode
+from .program_executor import ProgramExecutor
+
+
+class DatasetPlugin(ABC):
+    """
+    All dataset plugins derive from this class.
+    To register as a dataset they must decorate themselves with or call DatasetPlugin.register_plugin()
+    """
+
+    _executor: ProgramExecutor
+    # Context -> constructor_keys -> dataset plugin
+    _plugins: Dict[Context, Dict[set[str], DatasetPlugin]] = {}
+    _META_COLUMNS = ["run_id"]
+
+    def __init__(
+        self,
+        name: str,
+        logical_key: Optional[str] = None,
+        columns: Optional[Union[Iterable[str], str]] = None,
+        run_id: Optional[str] = None,
+        mode: Union[Mode, str] = Mode.READ,
+        class_field_name: Optional[str] = None,
+    ):
+        """
+
+        :param name: The dataset logical name.
+        :param logical_key:
+            The logical primary key, strongly suggested, and can later be
+            used when creating Hive/Dynamo tables or registering with a Catalog.
+        :param columns: Fetch columns
+        :param run_id: The ML Program run_id partition to select from.
+        :param mode: The data access read/write mode
+        :param class_field_name:
+            To be used by @dataset decorator to set the class
+            self.`class_field_name` field name, otherwise sets self.`name` as the class name
+        """
+        self.name = name
```

Comment:
so @sruthinagalla stumbled upon this: currently we have use cases where our constructor key that maps to `name` or `class_field_name` in our plugin has dashes in it.

```python
@dataset(name="foo-bar")
def method(self):
    self.foo-bar.read() <-- obviously is not valid
```

It is accessible through the objects __dict__ attribute.

```python
@dataset(name="foo-bar")
def method(self):
    self.__dict__["foo-bar"].read()
```

it seems like it would be most appropriate to inspect the `name` and `class_field_name` here and either raise and exception to avoid the above from happening, or adjusting the string value and letting the user know.

Reply:
This is a good catch.
I'm adding the following (with unit tests):
```python
            if dataset._class_field_name.isidentifier() or keyword.iskeyword(dataset._class_field_name):
                if class_field_name is None:
                    raise ValueError(f"{name} is not a valid Python identifier, "
                                     "please use 'class_field_name' for class field name")
                else:
                    raise ValueError(f"{class_field_name} is not a valid Python identifier")

            setattr(self, dataset._class_field_name, dataset)
```

to https://github.com/zillow/datasets/blob/ed6d0767a0f822461a8af4fc51411c9cbbc583b6/datasets/datasets_decorator.py#L13-L15

Reply:
With further discussion, if we enforce dataset name to **Upper Camel Casing**, then we can automatically create a `field_name` for you.

Reply:
https://github.com/zillow/datasets/pull/9